pinot-server/src/main/java/org/apache/pinot/server/starter/helix/HelixInstanceDataManager.java

```diff
@@ -237,8 +237,13 @@ private void reloadSegment(@Nonnull String tableNameWithType, @Nonnull SegmentMe
       // Delete segment temporary directory
       FileUtils.deleteDirectory(segmentTempDir);
     } finally {
-      LoaderUtils.reloadFailureRecovery(indexDir);
-      segmentLock.unlock();
+      try {
+        LoaderUtils.reloadFailureRecovery(indexDir);
+      } catch (Exception e) {
```

Comment:
Please  remove this exception block, and let it throw exception. Try/finally should be fine.
thanks

Reply:
This exception can override the current exception, which is not desired.
I would recommend moving `LoaderUtils.reloadFailureRecovery(indexDir);` after line 194 (recover the segment before trying to reload it)

Reply:
To avoid overriding, I chose to just log the recovery failure exception. Now the recovery is done only during catch block and we retain the top level exception (due to reload failure), log the recovery failure exception (if any) and also add it as a suppressed one to the parent exception. 

Lock is released in finally and we still throw the top level exception.